File: 0401-binary-watch/0401-binary-watch.py
class Solution:
    def readBinaryWatch(self, turnedOn: int) -> List[str]:
        ans = []
        def dfs(pos, count, h, m):
            if h > 11 or m > 59: return

            if pos == 10:
                if count == turnedOn:
                    ans.append(f"{h}:{m:02d}")
                return

            # off
            dfs(pos + 1, count, h, m)

            # on
            if pos < 4:
                dfs(pos + 1, count + 1, h + (1 << pos), m)
            else:
                dfs(pos + 1, count + 1, h, m + (1 << (pos - 4)))

        
        dfs(0, 0, 0, 0)
        return ans

File: 0401-binary-watch/0401-binary-watch.go
func readBinaryWatch(turnedOn int) []string {
	ans := []string{}

	var dfs func(pos, countOn, h, m int)
	dfs = func(pos, countOn, h, m int) {
        if h > 11 || m > 59 { return }

        if pos == 10 {
            if countOn == turnedOn {
                ans = append(ans, fmt.Sprintf("%d:%02d", h, m))
            }
            return
        }

        // LED off
        dfs(pos+1, countOn, h, m)

        // LED on
        if pos < 4 {
            dfs(pos+1, countOn+1, h+(1<<pos), m)
        } else {
            dfs(pos+1, countOn+1, h, m+(1<<(pos-4)))
        }
	}

    dfs(0, 0, 0, 0)
    return ans
}

// backtracking
// time: O(1)
// space: O(1)